File: lesson17-static-vs-dynamic-dispatch/Truncate.java
class Truncate {
    protected final double nr;

    public Truncate(double nr) { this.nr = nr; }

    // open by default => virtual
    public double cut() { return (int) nr; }
    public double cut(int decimals) {
        int d = 1;
        while(decimals-- > 0) d *= 10;
        return ((double) ((int) (nr * d))) / d;
    }
}

class Round extends Truncate {

    public Round(double nr) { super(nr); }

    // override by default
    public double cut() { return (int) (nr + 0.5); }
    public double cut(int decimals) {
        int d = 1;
        while (decimals-- > 0) { d *= 10; }
        return ((int) ((nr * d) + 0.5)) / (double) d;
    }
}

File: lesson17-static-vs-dynamic-dispatch/Truncate.kt
open class Truncate(protected val nr: Double) {
    open fun cut(): Double = nr.toInt().toDouble()

    open fun cut(decimals: Int): Double {
        var decimals = decimals
        var d = 1
        while (decimals-- > 0) { d *= 10 }
        return (nr * d).toInt().toDouble() / d
    }
}

class Round(nr: Double) : Truncate(nr) {
    override fun cut(): Double = ((nr + 0.5).toInt()).toDouble()

    override fun cut(decimals: Int): Double {
        var decimals = decimals
        var d = 1
        while (decimals-- > 0) { d *= 10 }
        return (nr * d + 0.5).toInt() / d.toDouble()
    }
}
